neural_compressor/adaptor/torch_utils/awq.py

```diff
@@ -0,0 +1,435 @@
+          weight_config[k]['bits'] == -1):
+            skip_op_set.add(k)
+    for k in skip_op_set:
+        if k in absorb_to_layer:
+            absorb_to_layer.pop(k)
+        if k in no_absorb_layers:
+            no_absorb_layers.remove(k)
+    if len(skip_op_set) > 0:
+        logger.info(f"{skip_op_set} are skipped when running AWQ optimization")
+
+    block_prefix, block_num = get_block_prefix(model)
+    for i in range(block_num):
+        block_absorb_dict[i] = []
+        block_name = block_prefix + '.' + str(i) + '.'
+        for k, v in absorb_to_layer.items():
+            name_list =tuple(vv for vv in v if block_name in vv)
+            if len(name_list) > 0:
+                block_absorb_dict[i].append(name_list)
+                absorb_layer_dict[name_list] = k
+        if not folding:
+            for k in no_absorb_layers:
+                if block_name in k:
+                    name_list = tuple([k])
+                    block_absorb_dict[i].append(name_list)
+                    absorb_layer_dict[name_list] = k
+    logger.debug(f"The absorbed layers per block: {block_absorb_dict}")
+    logger.debug(f"The absorb_layer_dict: {absorb_layer_dict}")
+    return block_absorb_dict, absorb_layer_dict
+
+
+@torch.no_grad()
+def _get_weight_scale(weight, q_group_size=-1):
+    org_shape = weight.shape
+    if q_group_size > 0:
+        weight = weight.view(-1, q_group_size)
+    scale = weight.abs() / weight.abs().amax(dim=1, keepdim=True)
+    scale = scale.view(org_shape)
+    scale = scale.mean(0)
+    return scale
+
+
+@torch.no_grad()
+def _get_act_scale(input_val):
+    tmp = [x.abs().view(-1, x.shape[-1]) for x in input_val]
+    tmp = torch.cat(tmp, dim=0)
+    return tmp.mean(0)
+
+
+class ActAwareWeightQuant:
+    """Implementation of Activation-aware Weight quantization (AWQ) algo."""
+    def __init__(self, model, example_inputs=None, calib_func=None, dataloader=None, n_samples=128,
+                 bits=4,  group_size=32, scheme='asym', sym_full_range=False, weight_config={},):
+        self.example_inputs = example_inputs
+        if example_inputs is None:
+            assert dataloader is not None, "datalaoder or example_inputs is required."
+            self.example_inputs = get_example_input(dataloader)
+        # Step 1: get hidden states and kwargs of first block. 
+        self.total_block_args, self.total_block_kwargs = get_hidden_states(
+            model, dataloader=dataloader, n_samples=n_samples, calib_func=calib_func
+        )
+        # Step 2: get block list and block prefix, number
+        self.block_prefix, self.block_num = get_block_prefix(model)
+        self.block_list = fetch_module(model, self.block_prefix)
+        self.bits = bits
+        self.group_size = group_size
+        self.scheme = scheme
+        self.sym_full_range = sym_full_range
+        self.weight_config = weight_config
+        self.model = model
```

Comment:
you should call model.eval if it is not in eval mode

Reply:
ok